src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,399 @@
+        {
+            var res = DoubleParser.Parse(out value, memory);
+            Contracts.Assert(res != DoubleParser.Result.Empty || value == 0);
+            return res <= DoubleParser.Result.Empty;
+        }
+
+        /// <summary>
+        /// This produces zero for an empty string.
+        /// </summary>
+        public static bool TryParse(out Double value, ReadOnlyMemory<char> memory)
+        {
+            var res = DoubleParser.Parse(out value, memory);
+            Contracts.Assert(res != DoubleParser.Result.Empty || value == 0);
+            return res <= DoubleParser.Result.Empty;
+        }
+
+        public static uint Hash(uint seed, ReadOnlyMemory<char> memory) => Hashing.MurmurHash(seed, memory);
+
+        public static NormStr AddToPool(NormStr.Pool pool, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(pool, nameof(pool));
+            return pool.Add(memory);
+        }
+
+        public static NormStr FindInPool(NormStr.Pool pool, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(pool, nameof(pool));
+            return pool.Get(memory);
+        }
+
+        public static void AddToStringBuilder(StringBuilder sb, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(sb, nameof(sb));
+            if (!memory.IsEmpty)
+                sb.AppendAll(memory);
+        }
+
+        public static void AddLowerCaseToStringBuilder(StringBuilder sb, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(sb, nameof(sb));
+
+            if (!memory.IsEmpty)
+            {
+                int ichLim = memory.Length;
+                int min = 0;
+                int j;
+                for (j = min; j < ichLim; j++)
+                {
+                    char ch = CharUtils.ToLowerInvariant(memory.Span[j]);
+                    if (ch != memory.Span[j])
+                    {
+                        sb.Append(memory, min, j - min).Append(ch);
+                        min = j + 1;
+                    }
+                }
+
+                Contracts.Assert(j == ichLim);
+                if (min != j)
+                    sb.Append(memory, min, j - min);
+            }
+        }
+
+        // REVIEW: Can this be faster?
+        private static bool ContainsChar(char ch, char[] rgch)
+        {
+            Contracts.CheckNonEmpty(rgch, nameof(rgch));
+
+            for (int i = 0; i < rgch.Length; i++)
+            {
+                if (rgch[i] == ch)
+                    return true;
+            }
+            return false;
+        }
+
+        public static StringBuilder AppendAll(this StringBuilder sb, ReadOnlyMemory<char> memory) => Append(sb, memory, 0, memory.Length);
+
+        public static StringBuilder Append(this StringBuilder sb, ReadOnlyMemory<char> memory, int startIndex, int length)
+        {
+            Contracts.Assert(startIndex >= 0, nameof(startIndex));
```

Comment:
>Assert [](start = 22, length = 6)

In contrast, this, on the other hand, is a public method in a public class, and so a check is appropriate, rather than the assert written here.